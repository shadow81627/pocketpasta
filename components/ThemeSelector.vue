<template>
  <client-only>
    <b-form-select
      id="theme"
      v-model="currentTheme"
      name="theme"
      aria-label="theme"
    >
      <template slot="first">
        <option disabled value="">Please select a theme</option>
      </template>
      <option
        v-for="theme in $store.state.themes"
        :key="theme.value"
        :value="theme.value"
      >
        {{ capitalize(theme.value) }}
      </option>
    </b-form-select>
  </client-only>
</template>

<script>
export default {
  computed: {
    currentTheme: {
      get() {
        return this.$colorMode.preference;
      },
      set(value) {
        this.$colorMode.preference = value;
        // timeout required to wait for color mode to be set.
        setTimeout(
          () =>
            (this.$vuetify.theme.dark = this.$store.getters.getThemeById(
              this.$colorMode.value,
            ).dark),
          100,
        );
      },
    },
    themeMeta() {
      return this.$store.state.themes.map((theme) => {
        const themeLink = {
          hid: `theme-${theme.value}`,
          href: theme.href,
          rel: 'stylesheet',
          as: 'style',
          skip:
            this.$store.getters.getThemeById(theme.value).href === '' ||
            !this.$store.getters.getThemeById(theme.value).href,
        };
        if (theme.value !== this.currentTheme) {
          themeLink.disabled = true;
          themeLink.rel = 'preload';
        }
        return themeLink;
      });
    },
  },
  methods: {
    capitalize(value) {
      if (!value) return '';
      value = value.toString();
      return value.charAt(0).toUpperCase() + value.slice(1);
    },
  },
  head() {
    return {
      link: [].concat(this.themeMeta),
    };
  },
};
</script>
