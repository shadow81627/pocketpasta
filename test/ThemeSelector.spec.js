import ThemeSelector from '@/components/ThemeSelector.vue';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import BootstrapVuePlugin from 'bootstrap-vue';

const localVue = createLocalVue();

localVue.use(Vuetify);
localVue.use(BootstrapVuePlugin);
localVue.use(Vuex);

describe('ThemeSelector', () => {
  let getters;
  let state;
  let store;
  let mutations;

  let vuetify;

  beforeEach(() => {
    vuetify = new Vuetify({
      mocks: {
        $vuetify: {
          theme: {},
        },
      },
    });
    state = {
      themes: [
        {
          value: 'darky',
          href: 'https://bootswatch.com/4/darkly/bootstrap.min.css',
        },
        {
          value: 'united',
          href: 'https://bootswatch.com/4/united/bootstrap.min.css',
        },
      ],
      theme: {
        value: 'darky',
        href: 'https://bootswatch.com/4/darkly/bootstrap.min.css',
      },
    };

    getters = {
      clicks: () => 2,
      inputValue: () => 'input',
      getThemeById: (state) => (value) =>
        state.themes.find((theme) => theme.value === value),
    };

    mutations = {
      setTheme(currentState, theme) {
        currentState.theme = theme;
      },
    };

    store = new Vuex.Store({
      state,
      getters,
      mutations,
    });
  });

  const factory = () =>
    shallowMount(ThemeSelector, {
      store,
      vuetify,
      localVue,
    });

  test('mounts properly', () => {
    const wrapper = factory();
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test('renders properly', () => {
    const wrapper = factory();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('theme getter', () => {
    const wrapper = factory();
    wrapper.vm.currentTheme = 'united';
    expect(wrapper.vm.currentTheme).toBeTruthy();
  });

  test('theme meta', () => {
    const wrapper = factory();
    expect(wrapper.vm.themeMeta).toBeTruthy();
  });

  test('head', () => {
    ThemeSelector.$store = store;
    expect(ThemeSelector.head()).toBeTruthy();
  });
});
