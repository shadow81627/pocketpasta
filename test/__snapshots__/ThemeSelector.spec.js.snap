// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSelector renders properly 1`] = `
"<client-only><b-form-select-stub id=\\"theme\\" name=\\"theme\\" options=\\"\\" valuefield=\\"value\\" textfield=\\"text\\" htmlfield=\\"html\\" disabledfield=\\"disabled\\" value=\\"darky\\" selectsize=\\"0\\" aria-label=\\"theme\\"><template></template> <option value=\\"darky\\">
      darky
    </option><option value=\\"united\\">
      united
    </option></b-form-select-stub></client-only>"
`;
